plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,422 @@
+package application_insights
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"time"
+	"unsafe"
+
+	"github.com/Microsoft/ApplicationInsights-Go/appinsights"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+type TelemetryTransmitter interface {
+	Track(appinsights.Telemetry)
+	Close() <-chan struct{}
+}
+
+type DiagnosticsMessageSubscriber interface {
+	Subscribe(appinsights.DiagnosticsMessageHandler) appinsights.DiagnosticsMessageListener
+}
+
+type ApplicationInsights struct {
+	InstrumentationKey      string
+	Timeout                 internal.Duration
+	EnableDiagnosticLogging bool
+	ContextTagSources       map[string]string
+	diagMsgSubscriber       DiagnosticsMessageSubscriber
+	transmitter             TelemetryTransmitter
+	diagMsgListener         appinsights.DiagnosticsMessageListener
+}
+
+var (
+	sampleConfig = `
+## Instrumentation key of the Application Insights resource.
+instrumentationKey = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx"
+
+## Timeout on close. If not provided, will default to 5s. 0s means no timeout (not recommended).
+# timeout = "5s"
+
+## Determines whether diagnostic logging (for Application Insights endpoint traffic) is enabled. Default is false.
+# enableDiagnosticLogging = "true"
+
+## ContextTagSources dictionary instructs the Application Insights plugin to set Application Insights context tags using metric properties.
+## In this dictionary keys are Application Insights context tags to set, and values are names of metric properties to use as source of data.
+## For example:
+# [outputs.application_insights.contextTagSources]
+# "ai.cloud.role" = "kubernetes_container_name"
+# "ai.cloud.roleInstance" = "kubernetes_pod_name"
+## will set the ai.cloud.role context tag to the value of kubernetes_container_name property (if present), 
+## and the ai.cloud.roleInstance context tag to the value of kubernetes_pod_name property.
+## For list of all context tag keys see https://github.com/Microsoft/ApplicationInsights-Go/blob/master/appinsights/contracts/contexttagkeys.go
+`
+	is32Bit        bool
+	is32BitChecked bool
+)
+
+func (a *ApplicationInsights) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *ApplicationInsights) Description() string {
+	return "Send telegraf metrics to Azure Application Insights"
+}
+
+func (a *ApplicationInsights) Connect() error {
+	if a.transmitter == nil && a.InstrumentationKey != "" {
+		a.transmitter = NewAppinsightsTransmitter(a.InstrumentationKey)
+	}
+
+	if a.EnableDiagnosticLogging && a.diagMsgSubscriber != nil {
+		a.diagMsgListener = a.diagMsgSubscriber.Subscribe(func(msg string) error {
+			logOutputMsg("%s", msg)
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 || a.transmitter == nil || a.InstrumentationKey == "" {
```

Comment:
Will remove the metric length test.

Regarding missing instrumentation key, the standard procedure for Application Insights is to no-op if the key is missing, but otherwise continue with the work. I will emit a warning, but I am pretty sure I do not want to stop Telegraf just because AppInsights instrumentation key is missing.

Reply:
Two reasons:

1. There might be other outputs that are functional and sending data somewhere
1. If Telegraf is used as a basis for metrics sidecar container in a Kubernetes environment, the failure of metrics container will cause the whole pod to be considered Failed and subject to relocation & restarts, affecting the primary pod functionality. IMO it is better to just warn about invalid configuration, but let the main container do its job, because that is the most important thing. Lack of metrics from such pod will be, hopefully, noticed fairly quickly, but the primary functionality will not be affected.

Reply:
I don't think we should support having the plugin both enabled and non-functional, it will just cause confusion and can be resolved before runtime, it should be easy enough to just not enable the plugin if no output is desired.  Also, this is contrary to how all the other outputs work so I don't want to introduce a different behavior.

It might be helpful to use the `--config-directory` to more easily enable/disable plugins.